tcp-server/src/main.rs

```diff
@@ -50,47 +45,53 @@ impl VecDequeMailbox {
 
 /// General questions:
 /// * how to work with `dyn Mailbox ` in multithreaded environment?
-fn main() -> io::Result<()> {
-    spawn_tcp_listener();
-    let client_thread = spawn_monitoring_thread();
-    client_thread.join().unwrap();
+#[tokio::main]
+async fn main() -> io::Result<()> {
+    tokio::join!(spawn_monitoring_thread(), spawn_tcp_listener(),);
     Ok(())
 }
 
 /// Spawns TcpListener thread, connects to 127.0.0.1:8080 and spawns thread per connection
-fn spawn_tcp_listener() {
-    thread::spawn(|| {
+fn spawn_tcp_listener() -> JoinHandle<()> {
+    return tokio::spawn(async {
         let mailbox = Arc::new(VecDequeMailbox::new());
-        let listener = TcpListener::bind("127.0.0.1:8080").unwrap();
-        for stream in listener.incoming() {
-            // TODO is shadowing the variable name idiomatic?
+        let listener = TcpListener::bind("127.0.0.1:8080").await.unwrap();
+
+        loop {
+            let next = listener.accept().await.unwrap();
             let mailbox = mailbox.clone();
-            // TODO: confirm that dropping the JoinHandle does not do anything
-            thread::spawn(|| {
-                println!("Server has a new client {:?}", stream);
-                handle_client(stream.unwrap(), mailbox);
+            tokio::spawn(async {
+                handle_client(BufReader::new(next.0), mailbox).await;
             });
         }
     });
 }
 
 /// TODO how can I pass the mailbox, preferably just mailbox: Mailbox or at least mailbox: Arc<Box<dyn Mailbox>>?
-fn handle_client(mut client: TcpStream, mailbox: Arc<Box<VecDequeMailbox>>) -> Result<(), io::Error> {
+async fn handle_client(
+    mut tcp_stream: BufReader<TcpStream>,
+    mailbox: Arc<Box<VecDequeMailbox>>,
+) -> Result<(), io::Error> {
     loop {
         let mut str = String::new();
-        BufReader::new(&client).read_line(&mut str);
+        let ret = tcp_stream.read_line(&mut str).await;
+        if let Ok(0) = ret {
```

Comment:
Magic Number.

Reply:
This is arguable. In many best practices 0 and 1 are not considered magic numbers. Here 0 means that 0 bytes were read indicating an EOF.